### PR TITLE
Remove "registration" metaclasses

### PR DESCRIPTION
**Pull request by matthewturk**
## PR Summary

Nearly since the code was created, we have utilized metaclasses to "register"
objects.  For instance, we do this to populate all of the data objects that are
available on a given dataset, all the dataset types we know about, etc etc.

Starting in python 3.6, we can start using the `__init_subclass__` method to
initialize a subclass.  This is pretty handy, because it means that instead of
doing a metaclass, we just have to do this one single routine.

One of the upsides of this is that now we can use other base classes for our
data objects -- for instance, if we want to use `HasTraits` as a base class.
Plus, it flattens things and makes it a bit simpler.

## PR Checklist

- [x] pass `black --check yt/`
- [x] pass `isort . --check --diff`
- [x] pass `flake8 yt/`
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by matthewturk** [comment]
Having a single file would be a great idea.  I'll give a shot with that; I want to make sure that it doesn't get *re*-initialized every import, so I need to review what the import process does to module-declared variables.

File: yt/frontends/ramses/particle_handlers.py [hunk 1]
```diff
@@ -57,6 +44,14 @@ class ParticleFileHandler(metaclass=RAMSESParticleFileHandlerRegistry):
     )
     local_particle_count = None  # The number of particle in the domain
 
+    def __init_subclass__(cls, *args, **kwargs):
+        """
+        This ensures that when a new subclass is created, it's registered.
```
[the remark this answers is not in the document]
Comment (matthewturk) [reply yt/frontends/ramses/particle_handlers.py:49]:
```suggestion
        Registers a subclass at creation.
```

**Comment by matthewturk** [comment]
As a note, one thing I *don't* particularly want to do is build out a mixin that has a subclass init thing, etc etc.  But maybe that would be another viable tactic?

**Comment by neutrinoceros** [comment]
> One of the upsides of this is that now we can use other base classes for our
data objects -- for instance, if we want to use HasTraits as a base class.

naive question: you mean metaclasses here I suppose ? I would assume it'd be possible to have more than one but I never actually had a use for that... is there a strong restriction on this ?

> Plus, it flattens things and makes it a bit simpler.

no doubt this is better !

**Comment by matthewturk** [comment]
> naive question: you mean metaclasses here I suppose ? I would assume it'd be possible to have more than one but I never actually had a use for that... is there a strong restriction on this ?

Hmm, so what I ran into was that I couldn't add a metaclass while also inheriting.  I tried the naive thing of just changing the class def of `YTDataContainer` to `class YTDataContainer(traitlets.HasTraits, metaclass=...)` and it threw an error at me.

**Comment by matthewturk** [comment]
(Which maybe is because `HasTraits` has a metaclass of its own!)

**Comment by neutrinoceros** [comment]
> Hmm, so what I ran into was that I couldn't add a metaclass while also inheriting

oh right, this actually makes a lot of sense. I can imagine the __mro__ nightmare of having to debug a class that has a fresh metaclass AND inheritance. I understand why the stdlib devs wouldn't allow this :)

**Comment by neutrinoceros** [comment]
A note to other reviewers : because the first arg in `__init_subclass__` is `cls`, I initially thought that we needed to add a `@classmethod` decorator. In fact, the decorator is _implicit_, so this change is perfectly fine on this point :-)
For details, see https://www.python.org/dev/peps/pep-0487/#requiring-an-explicit-decorator-on-init-subclass

**Comment by matthewturk** [comment]
Looks like I need to fix the `Enzo` time series definition.

**Comment by matthewturk** [comment]
Some of the data object registries were really only useful locally; I've left their registry objects in those files.  But, I've moved the ones that get imported in a couple places to a specific location.  I believe this also fixes the cookbook issue.

**Comment by neutrinoceros** [comment]
So there's exactly ONE test that fails, and seemingly on MacOS only (`yt.frontends.stream.tests.test_stream_unstructured.test_multi_mesh`); Of course it will require careful inspection, but I'm inclined to think that it actually shows the test itself is broken, and could probably be removed.

**Comment by neutrinoceros** [comment]
Even weirder: I can't reproduce this failure locally (works on my MacOS, ran through pytest or nose)

**Comment by matthewturk** [comment]
@neutrinoceros are you suggesting it's ... haunted??? :ghost:

**Comment by neutrinoceros** [comment]
Who you gonna call ?

**Comment by matthewturk** [comment]
I've merged with upstream, and hopefully this will fix travis.

**Pull request merged**